server/sonar-web/src/main/js/apps/marketplace/components/PluginActions.tsx

```diff
@@ -0,0 +1,166 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as React from 'react';
+import Checkbox from '../../../components/controls/Checkbox';
+import CheckIcon from '../../../components/icons-components/CheckIcon';
+import PluginUpdateButton from './PluginUpdateButton';
+import { Plugin, installPlugin, updatePlugin, uninstallPlugin } from '../../../api/plugins';
+import { isPluginAvailable, isPluginInstalled } from '../utils';
+import { translate } from '../../../helpers/l10n';
+
+interface Props {
+  plugin: Plugin;
+  refreshPending: () => void;
+}
+
+interface State {
+  acceptTerms: boolean;
+  loading: boolean;
+}
+
+export default class PluginActions extends React.PureComponent<Props, State> {
+  mounted: boolean;
+  state: State = { acceptTerms: false, loading: false };
+
+  componentDidMount() {
+    this.mounted = true;
+  }
+
+  componentWillUnmount() {
+    this.mounted = false;
+  }
+
+  doPluginAction = (apiAction: (data: { key: string }) => Promise<void | Response>) => {
+    this.setState({ loading: true });
+    apiAction({ key: this.props.plugin.key }).then(
+      () => {
+        this.props.refreshPending();
+        if (this.mounted) {
+          this.setState({ loading: false });
+        }
+      },
+      () => {
+        if (this.mounted) {
+          this.setState({ loading: false });
+        }
+      }
+    );
+  };
+
+  handleInstall = () => this.doPluginAction(installPlugin);
+  handleUpdate = () => this.doPluginAction(updatePlugin);
+  handleUninstall = () => this.doPluginAction(uninstallPlugin);
+  handleTermsCheck = (checked: boolean) => this.setState({ acceptTerms: checked });
+
+  render() {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') The function render has a cyclomatic complexity of 15 which is higher than the threshold of 10 [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)